lib/network/game_server.rb

```diff
@@ -16,11 +18,17 @@ def current_address
         end
 
         def start
-            open_connection
-            print_connection_info
-            accept_connection
-            communicate
-            close_connection
+            begin
+                open_connection
+                print_connection_info
+                accept_connection
+                @io.print_output("Player one join the game")
```

Comment:
here too; just "food for thought".

Reply:
Proof reading comment; hope it is helpful.
```suggestion
                @io.print_output("Player one joined the game")
```

Reply:
> This is a huge PR, looks like you put in a lot of work. The larger the PR the harder it is for reviewers to review. Overall I think it looks pretty good. The use of hand spun mocks/spys for testing the server and client was a good move. I am approving, and sprinkled in some comments. Nothing is blocking, so do not feel like you need to respond to the comments in anyway at this time.

I agree with all the comments you made in this PR. I'll follow your recommendations. Thank you very much!